### PR TITLE
refactor: split writing of long and short values

### LINKED ISSUE #41: review smart-card writing of short and long value.
i.e. do we want to separate writing of short and long values?

### PR DESCRIPTION
**Pull request by eventualbuddha**
This PR ended up doing a fair amount that would have been better done in separate commits and/or separate PRs, but this is what I have so let's see if it's reviewable.

Closes #41. This is the main thing this PR is intended to do. It does this by changing the structure of the data written to the smartcard so that the header is its own chunk, the short value is its own chunk (with a flag that indicates whether there is a long value), and the long value is in chunks after that. See the doc comment in `Card` for an explanation.

The code was restructured quite a bit too. The idea is that we maintain a cache of short/long values and metadata in the card and also their new "pending" values when calls to `write_{short,long}_value` happen (also for `write_protect`). Then all changes get pushed to the card at once with the private `_flush` method. In theory we could later change the API to all writing these pending values without flushing to the card, but that's out of scope for now (though there's an argument to be made that `write_short_and_long` should do exactly this). The primary reason to do this is because we needed to know which chunks needed re-writing for any given set of changes. For example, writing the short value only requires writing chunk 1. Initializing the card requires writing chunks 1, 2, & 3, and usually one or more of them again depending on the write. This is sub-optimal (see TODOs).

I changed the testing setup a fair amount in a way that removes most mocking. The downside is that I don't have good assurances right now that reads and writes aren't happening when they shouldn't be. This is something we should add (see TODOs). On the plus side the tests now run very fast because I stubbed out the `time.sleep` calls in tests to be no-ops.

And finally, I got tired of VS Code nagging me about installing various development-time tools like `pylint`, so I just gave in. The only one I hooked up in CI so far is `mypy`, though there aren't a lot of type annotations yet.

### TODOs

- [ ] add tests that ensure `read_chunk` and `write_chunk` aren't happening more than needed
- [ ] optimize number of writes when writing requires initialization?
- [x] don't gzip the long value if the result would be longer
- [ ] test with real smart cards and a working reader

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -22,5 +22,8 @@ test:
 coverage:
 	python3 -m pytest --cov=smartcards --cov-report term-missing --cov-fail-under=100 tests/
 
+typecheck:
```
Comment (benadida) [inline Makefile]: uhuh, I'm sure this was VsCode nagging ;)

File: tests/test_card.py [hunk 2]
```diff
@@ -1,110 +1,158 @@
...
+        encoded_long_value)
+    return card_header + card_short_value + card_long_value
 
-import pytest, json, os, secrets
 
 # data on card mocked
-CONTENT_BYTES = b'hello1234'
-CARD_BYTES = b'VX.'+bytes([0x01])+bytes([0x00])+bytes([len(CONTENT_BYTES), 0x00, 0x00]) + CONTENT_BYTES
+SHORT_VALUE: bytes = b'hello1234'
+LONG_VALUE: bytes = b'here is a long value I hope you like it'
```
Comment (benadida) [inline tests/test_card.py]: I do like it. It's nice.

File: tests/test_card.py [hunk 3]
```diff
@@ -1,110 +1,158 @@
...
+# same as default card, but write-protected
+WRITE_PROTECTED_CARD_BYTES = build_card_data(
+    SHORT_VALUE, None, write_protected=True)
+
+# card with short and long values
+CARD_BYTES_WITH_LONG_VALUE = build_card_data(
+    SHORT_VALUE, LONG_VALUE)
+
+
+def flip_byte(data, index):
```
Comment (benadida) [inline tests/test_card.py]: should we be testing with a single bit flip rather than a whole byte?

Comment (eventualbuddha) [reply tests/test_card.py]: Sure, should be easy enough.

**Comment by eventualbuddha** [comment]
@benadida I snuck in another change: we now only gzip things if it helps. Notably, for the existing ballot encoding, it likely doesn't.

File: Makefile (hunk 1 above)
Comment (eventualbuddha) [reply Makefile]: Well I had to add it to CI and you do love your makefiles so… 😝

**Comment by eventualbuddha** [comment]
Closing this as it's having issues with the real reader. Might be related to the bytes vs. bytearray thing @benadida identified. Much of this is already included in #46, #47, and #48. The rest will come later.

**Pull request closed without merging**